File: app/src/main/java/com/tebet/mojual/data/local/db/AppDbHelper.java
/*
 *  Copyright (C) 2017 MINDORKS NEXTGEN PRIVATE LIMITED
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      https://mindorks.com/license/apache-v2
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License
 */

package com.tebet.mojual.data.local.db;

import com.tebet.mojual.data.local.db.dao.BankDao;
import com.tebet.mojual.data.local.db.dao.CityDao;
import com.tebet.mojual.data.local.db.dao.UserProfileDao;
import com.tebet.mojual.data.models.Bank;
import com.tebet.mojual.data.models.City;
import com.tebet.mojual.data.models.UserProfile;
import io.reactivex.Observable;
import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.List;
import java.util.concurrent.Callable;


@Singleton
public class AppDbHelper implements DbHelper {

    private final AppDatabase mAppDatabase;

    @Inject
    public AppDbHelper(AppDatabase appDatabase) {
        this.mAppDatabase = appDatabase;
    }

    @Override
    public Observable<UserProfileDao> getUserProfile() {
        return Observable.fromCallable(new Callable<UserProfileDao>() {
            @Override
            public UserProfileDao call() throws Exception {
                return mAppDatabase.userProfileDao();
            }
        });
    }

    @Override
    public Observable<Boolean> insertUserProfile(UserProfile userProfile) {
        return Observable.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                mAppDatabase.userProfileDao().insertUserProfile(userProfile);
                return true;
            }
        });
    }

    @Override
    public Observable<BankDao> getBank() {
        return Observable.fromCallable(new Callable<BankDao>() {
            @Override
            public BankDao call() throws Exception {
                return mAppDatabase.bankDao();
            }
        });
    }

    @Override
    public Observable<Boolean> insertBank(Bank bank) {
        return Observable.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                mAppDatabase.bankDao().insertBank(bank);
                return true;
            }
        });
    }

    @Override
    public Observable<Boolean> insertBanks(@Nullable List<Bank> banks) {
        return Observable.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                mAppDatabase.bankDao().insertAll(banks);
                return true;
            }
        });
    }

    @Override
    public Observable<CityDao> getCity() {
        return Observable.fromCallable(new Callable<CityDao>() {
            @Override
            public CityDao call() throws Exception {
                return mAppDatabase.cityDao();
            }
        });
    }

    @Override
    public Observable<Boolean> insertCity(City city) {
        return Observable.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                mAppDatabase.cityDao().insertCity(city);
                return true;
            }
        });
    }
}

File: app/src/main/java/com/tebet/mojual/data/local/db/AppDbHelper.kt
package com.tebet.mojual.data.local.db

import com.tebet.mojual.data.local.db.dao.*
import com.tebet.mojual.data.models.*
import io.reactivex.Observable
import javax.inject.Inject
import javax.inject.Singleton


@Singleton
class AppDbHelper @Inject
constructor(private val mAppDatabase: AppDatabase) : DbHelper {

    override val userProfile: Observable<UserProfileDao>
        get() = Observable.fromCallable { mAppDatabase.userProfileDao() }

    override val bank: Observable<BankDao>
        get() = Observable.fromCallable { mAppDatabase.bankDao() }

    override val region: Observable<RegionDao>
        get() = Observable.fromCallable { mAppDatabase.regionDao() }

    override val city: Observable<CityDao>
        get() = Observable.fromCallable { mAppDatabase.cityDao() }

    override val asset: Observable<AssetDao>
        get() = Observable.fromCallable { mAppDatabase.assetDao() }

    override fun insertUserProfile(userProfile: UserProfile): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.userProfileDao().insertUserProfile(userProfile)
            true
        }
    }

    override fun insertBank(bank: Bank): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.bankDao().insertBank(bank)
            true
        }
    }

    override fun insertBanks(banks: List<Bank>): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.bankDao().insertAll(banks)
            true
        }
    }

    override fun insertRegions(regions: List<Region>): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.regionDao().insertAll(regions)
            true
        }
    }

    override fun insertCity(city: City): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.cityDao().insertCity(city)
            true
        }
    }

    override fun insertCities(cities: List<City>): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.cityDao().insertAll(cities)
            true
        }
    }

    override fun clearAllTables(): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.clearAllTables()
            true
        }
    }

    override fun clearAllProfiles(): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.userProfileDao().nukeTable()
            true
        }
    }

    override fun insertAssets(asset: List<Asset>): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.assetDao().insertAll(asset)
            true
        }
    }

    override fun clearAllAssets(): Observable<Boolean> {
        return Observable.fromCallable {
            mAppDatabase.assetDao().nukeTable()
            true
        }
    }

}
